File: app/server/notify/channels/telegram.ts
/**
 * Telegram Bot 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'telegram' });

export const telegramChannel: NotifyChannel = {
    name: 'telegram',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('TG_BOT_TOKEN', 'TG_USER_ID')) {
                resolve({ success: false, message: 'TG_BOT_TOKEN 或 TG_USER_ID 未配置' });
                return;
            }

            const TG_BOT_TOKEN = getConfig('TG_BOT_TOKEN');
            const TG_USER_ID = getConfig('TG_USER_ID');
            const TG_API_HOST = getConfig('TG_API_HOST') || 'https://api.telegram.org';

            const url = `${TG_API_HOST}/bot${TG_BOT_TOKEN}/sendMessage`;
            const content = `${text}\n\n${desp}`.substring(0, 4096);

            const body = {
                chat_id: TG_USER_ID,
                text: content,
                parse_mode: 'HTML'
            };

            $.post({ url, json: body }, (err, _res, resBody) => {
                if (err) {
                    logger.error({ err }, 'Telegram 发送失败');
                    resolve({ success: false, error: err as Error });
                    return;
                }

                const data = resBody as { ok?: boolean; description?: string };
                if (data?.ok) {
                    logger.info('Telegram 发送成功');
                    resolve({ success: true, message: '发送成功' });
                } else {
                    logger.warn({ body: resBody }, 'Telegram 发送失败');
                    resolve({ success: false, message: data?.description || '发送失败' });
                }
            });
        });
    }
};

// 检查是否启用
if (hasConfig('TG_BOT_TOKEN', 'TG_USER_ID')) {
    telegramChannel.enabled = true;
}

export default telegramChannel;

File: app/server/internal/notify/channels/telegram.go
package channels

import (
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type Telegram struct {
	enabled bool
}

func (c *Telegram) Name() string            { return "telegram" }
func (c *Telegram) Enabled() bool           { return c.enabled }
func (c *Telegram) SetEnabled(enabled bool) { c.enabled = enabled }

func (c *Telegram) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("TG_BOT_TOKEN", "TG_USER_ID") {
		return notify.NotifyResult{Success: false, Message: "TG_BOT_TOKEN 或 TG_USER_ID 未配置"}
	}

	tgBotToken := notify.GetConfig("TG_BOT_TOKEN")
	tgUserID := notify.GetConfig("TG_USER_ID")
	tgAPIHost := notify.GetConfig("TG_API_HOST")
	if tgAPIHost == "" {
		tgAPIHost = "https://api.telegram.org"
	}

	url := tgAPIHost + "/bot" + tgBotToken + "/sendMessage"

	content := text + "\n\n" + desp
	if len(content) > 4096 {
		content = content[:4096]
	}

	body := map[string]string{
		"chat_id":    tgUserID,
		"text":       content,
		"parse_mode": "HTML",
	}

	resp, err := notify.HTTPPost(url, notify.HttpRequestOptions{JSON: body})
	if err != nil {
		slog.Error("Telegram 发送失败", "err", err)
		return notify.NotifyResult{Success: false, Message: fmt.Sprintf("发送失败: %v", err), Err: err}
	}

	var result struct {
		Ok          bool   `json:"ok"`
		Description string `json:"description"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &result); err != nil {
		slog.Error("Telegram 解析响应失败", "err", err)
		return notify.NotifyResult{Success: false, Message: "解析响应失败", Err: err}
	}

	if result.Ok {
		slog.Info("Telegram 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	msg := result.Description
	if msg == "" {
		msg = "发送失败"
	}
	slog.Warn("Telegram 发送失败", "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: msg}
}

func init() {
	ch := &Telegram{}
	if notify.HasConfig("TG_BOT_TOKEN", "TG_USER_ID") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
